packages/portal/backend/src/server/common/ClasslistAgent.ts

```diff
@@ -0,0 +1,150 @@
+        Log.info("IntegrationController::fetchClasslist - start");
+        try {
+            const uri = this.getClasslistUri();
+            const options = {
+                headers: {
+                    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:40.0) Gecko/20100101 Firefox/40.0' // for testing
+                },
+                uri,
+                rejectUnauthorized: false
+            };
+            return JSON.parse(await rp(options));
+        } catch (err) {
+            Log.error("IntegrationController::fetchClasslist - ERROR: " + err);
+            throw new Error("Could not fetch Classlist " + err.message);
+        }
+    }
+
+    private getClasslistUri() {
+        const config = Config.getInstance();
+        const auth = config.getProp(ConfigKey.classlist_username) + ':' + config.getProp(ConfigKey.classlist_password);
+        const uri = config.getProp(ConfigKey.classlist_uri);
+
+        if (uri.indexOf('https://') === 0) {
+            return 'https://' + auth + '@' + uri.slice(8);
+        } else {
+            throw new Error('https:// protocol is required for API integration');
+        }
+    }
+
+    public async processClasslist(personId: string = null, path: string = null,  data: any): Promise<Person[]> {
+        Log.trace("PersonController::processClasslist(...) - start");
+
+        if (path !== null) {
+            data = await new CSVParser().parsePath(path);
+        }
+
+        this.duplicateDataCheck(data, ['ACCT', 'CWL']);
+        this.missingDataCheck(data, ['ACCT', 'CWL']);
+        const peoplePromises: Array<Promise<Person>> = [];
+
+        for (const row of data) {
+            // Log.trace(JSON.stringify(row));
+            if (typeof row.ACCT !== 'undefined' && typeof row.CWL !== 'undefined' &&
+                typeof row.SNUM !== 'undefined' && typeof row.FIRST !== 'undefined' &&
+                typeof row.LAST !== 'undefined' && typeof row.LAB !== 'undefined') {
+                const p: Person = {
+                    id:            row.ACCT.toLowerCase(), // id is CSID since this cannot be changed
+                    csId:          row.ACCT.toLowerCase(),
+                    // github.ugrad.cs wanted row.ACCT; github.students.cs and github.ubc want row.CWL
+                    githubId:      row.CWL.toLowerCase(),
+                    studentNumber: row.SNUM,
+                    fName:         row.FIRST,
+                    lName:         row.LAST,
+
+                    kind:   PersonKind.STUDENT,
+                    URL:    null,
+                    labId:  row.LAB,
+                    custom: {}
+                };
+                peoplePromises.push(this.pc.createPerson(p));
+            } else {
+                Log.error('PersonController::processClasslist(..) - column missing from: ' + JSON.stringify(row));
+                peoplePromises.push(Promise.reject('Required column missing (required: ACCT, CWL, SNUM, FIRST, LAST, LAB).'));
```

Comment:
This was probably my code, but I would check to make sure it is behaving as expected. I don't know that the `Promise.reject` would need to be pushed on `peoplePromises`. I think just `return Promise.reject(..)` might be enough?

Reply:
Hi Reid,

This is your code, which has worked in the past-- and reliably. The intended business logic is not clear though. 

From integrating the Classlist API, what we found is that if a student had an `undefined` LAB column because they did not register for a class, the rejection would trigger an error message for that failed student entry. However, most students had a defined LAB property. In this case, I believe that almost all of the students were still entered into the database. This was different behaviour than when we parsed the CSV sheet.

In the case of the CSV sheet, if the LAB column is defined in the CSV, we will never run into this rejection error because, even if the entry has an empty string, every parsed entry will have a LAB property due to how a LAB property is assumed for each CSV row by the parser. It is not possible to hit this rejection code with a CSV unless a CSV column is not defined, in which case all of the entries must fail. If a rejection were to occur, without all of the entries failing, my guess is that it would be when `this.pc.createPerson(p)` fails because of a database constraint (if any constraints exist at the moment in the data layer).

This is probably a very surprising answer, but overall, this code has worked really reliably from what I've seen when testing it. If we need it to do something specific for both CSV and Classlist API data, we can make it happen. I will try create some test cases for this on Monday if you have more questions about it.

Reply:
I don't know if we can just return Promise.reject() because of the `for loop` that the return would occur in. We could throw an error and short-circuit the entire classlist entry operation though. That would make everything fail if one user entry fails.

Reply:
@rtholmes We are going ahead with the merge request, but we are opening a separate issue where we can discuss the business logic.

Reply:
See #260